include/util/static_graph.hpp

```diff
@@ -18,54 +20,87 @@ namespace osrm
+using NodeIterator = NodeID;
+using EdgeIterator = NodeID;
+
+struct NodeArrayEntry
+{
+    // index of the first edge
+    EdgeIterator first_edge;
+};
+
+template <typename EdgeDataT> struct EdgeArrayEntry;
+
+template <typename EdgeDataT> struct EdgeArrayEntry
+{
+    NodeID target;
+    EdgeDataT data;
+};
+
+template <> struct EdgeArrayEntry<void>
+{
+    NodeID target;
+};
+
+template <typename EdgeDataT> struct SortableEdgeWithData;
+
+template <> struct SortableEdgeWithData<void>
+{
+    NodeIterator source;
+    NodeIterator target;
+
+    SortableEdgeWithData(NodeIterator source, NodeIterator target) : source(source), target(target)
     {
-      public:
-        NodeIterator source;
-        NodeIterator target;
-        EdgeDataT data;
-
-        template <typename... Ts>
-        InputEdge(NodeIterator source, NodeIterator target, Ts &&... data)
-            : source(source), target(target), data(std::forward<Ts>(data)...)
-        {
-        }
-        bool operator<(const InputEdge &right) const
+    }
+
+    bool operator<(const SortableEdgeWithData &right) const
+    {
+        if (source != right.source)
         {
-            if (source != right.source)
-            {
-                return source < right.source;
-            }
-            return target < right.target;
+            return source < right.source;
         }
-    };
+        return target < right.target;
+    }
+};
 
-    struct NodeArrayEntry
-    {
-        // index of the first edge
-        EdgeIterator first_edge;
-    };
+template <typename EdgeDataT> struct SortableEdgeWithData : SortableEdgeWithData<void>
+{
+    using Base = SortableEdgeWithData<void>;
+
+    EdgeDataT data;
 
-    struct EdgeArrayEntry
+    template <typename... Ts>
+    SortableEdgeWithData(NodeIterator source, NodeIterator target, Ts &&... data)
+        : Base{source, target}, data{std::forward<Ts>(data)...}
     {
-        NodeID target;
-        EdgeDataT data;
-    };
+    }
+};
+
+} // namespace static_graph_details
+
+template <typename NodeT, typename EdgeT, bool UseSharedMemory = false> class FlexibleStaticGraph
+{
+    static_assert(traits::HasFirstEdgeMember<NodeT>::value,
+                  "Model for compatible Node type requires .first_edge member attribute");
+    static_assert(traits::HasTargetMember<EdgeT>::value,
+                  "Model for compatible Node type requires .target member attribute");
+
+  public:
+    using NodeIterator = static_graph_details::NodeIterator;
+    using EdgeIterator = static_graph_details::EdgeIterator;
+    using EdgeRange = range<EdgeIterator>;
+    using NodeArrayEntry = NodeT;
+    using EdgeArrayEntry = EdgeT;
 
     EdgeRange GetAdjacentEdgeRange(const NodeID node) const
     {
         return irange(BeginEdges(node), EndEdges(node));
     }
 
-    template <typename ContainerT> StaticGraph(const int nodes, const ContainerT &graph)
+    template <typename ContainerT> FlexibleStaticGraph(const int nodes, const ContainerT &graph)
```

Comment:
we could think about reverting these changes completely, since the static graph is no longer used in partitioning and we don't need the adjustments anymore. @daniel-j-h since you added the traits, whats your take?